R/sql_methods.R

```diff
@@ -0,0 +1,170 @@
+#'
+#' Function to_sql(df, name, dbname, schemaname, if_exists, index, index_label, nanoseconds).
+#' Creates table in the connected nuvolos schema from an R dataframe.
+#' On Nuvolos the database and schema are by default the ones the user is working in, from local machine they need to be specified.
+#' The function supports bulk loading.
+#' 
+#' @param df Name of the R dataframe to be written to a table.
+#' @param name The name of the database table. It will only be quoted and case sensitive if it contains keywords or special chars
+#' @param dbname The name of the database to which data will be inserted.
+#' @param schemaname The name of the schema to which data will be inserted.
+#' @param if_exists How to behave if the table already exists. {‘fail’, ‘replace’, ‘append’}, default ‘fail’
+#' \itemize{
+#' \item fail: Raise a ValueError.
+#' \item replace: Drop the table before inserting new values.
+#' \item append: Insert new values to the existing table.
+#' }
+#' @param index bool, default True: Write DataFrame index as a column. Uses index_label as the column name in the table.
+#' @param index_label column label for index column(s). If None is given (default) and index is True, then the index names are used. A sequence should be given if the DataFrame uses MultiIndex.
+#' @param nanoseconds if True, nanosecond timestamps will be used to upload the data. Limits timestamp range from 1677-09-21 00:12:43.145224192 to 2262-04-11 23:47:16.854775807.
+#' @return Returns the COPY INTO command's results to verify ingestion in the form of a tuple of whether all chunks were ingested correctly, # of chunks, # of ingested rows, and ingest's output.
+#'
+#' @examples
+#' to_sql(df = df, name = "table", if_exists = 'replace', index = FALSE)
+#' to_sql(df = df, name = "table", dbname = "space_1", schemaname = "test_schema", if_exists = 'replace', index = FALSE)
+#'
+#' @export
+to_sql <- function(df,
+                   name,
+                   dbname=NULL,
+                   schemaname=NULL,
+                   if_exists="fail",
+                   index=TRUE,
+                   index_label=NULL,
+                   nanoseconds=FALSE){
+
+  require(reticulate)
+
+  # importing necessary python packages, installing if not available
+  nuvolos <- tryCatch({
+    import("nuvolos")
+  }, error = function(e){
+    py_install("nuvolos", pip = TRUE)
+    return(import("nuvolos"))
+  })
+  
+  username <- NULL
+  password <- NULL
+
+  # reading credentials for establishing connection
+  if (is_local()){
+    conn_param = get_local_info(username, password, dbname, schemaname)
+  } else {
+    conn_param = get_nuvolos_info(username, password, dbname, schemaname)
+  }
+  
+  # establishing connection with python-based nuvolos connector
+  con <- nuvolos$get_connection(username = username,
+                                password = password,
+                                dbname = dbname,
+                                schemaname = schemaname)
```

Comment:
The SQLAlchemy documentation doesn't mandate or emphasize it. The Connection.close will only release the connection back to the connection pool, it will probably not close the underlying Snowflake DB connection.
https://docs.sqlalchemy.org/en/14/core/connections.html#sqlalchemy.engine.Connection.close
Since the nuvolos.get_connection doesn't return the engine holding the pool with the connections, closing the connection won't make much of a difference.